protocol/0083-RFPR-on_chain_referral_program.md

```diff
@@ -0,0 +1,402 @@
+- `window_length`:  the number of epochs over which to evaluate a referral sets running volume
+
+```protobuf
+message UpdateReferralProgram{
+    changes: UpdateReferralProgramConfiguration{
+        benefit_tiers: [
+            {
+                "minimum_running_taker_volume": 10000,
+                "minimum_epochs": 0,
+                "referral_reward_factor": 0.001,
+                "referral_discount_factor": 0.001,
+            },
+            {
+                "minimum_running_taker_volume": 20000,
+                "minimum_epochs": 7,
+                "referral_reward_factor": 0.005,
+                "referral_discount_factor": 0.005,
+            },
+            {
+                "minimum_running_taker_volume": 30000,
+                "minimum_epochs": 31,
+                "referral_reward_factor": 0.010,
+                "referral_discount_factor": 0.010,
+            },
+        ],
+        closing_timestamp: 123456789,
+        window_length: 7,
+    }
+}
+```
+
+When submitting a referral program proposal through governance the following conditions apply:
+
+- a proposer cannot set an `closing_timestamp` less than the proposals `enactment_time`.
+- the number of tiers in `benefit_tiers` must be less than or equal to the network parameter `referralProgram.maxBenefitTiers`.
+- all `minimum_epochs` values must be an integer strictly greater than 0
+- all `referral_reward_factor` values must be greater than or equal to `0` and less than or equal to the network parameter `referralProgram.maxReferralRewardFactor`.
+- all `referral_discount_factor` values must be greater than or equal to `0` and be less than or equal to the network parameter `referralProgram.maxReferralDiscountFactor`.
+- `window_length` must be an integer strictly greater than zero.
+
+The referral program will start the epoch after the `enactment_timestamp` is reached.
+
+## Referral program lifecycle
+
+After a referral program [proposal](#governance-proposals) is validated and accepted by the network, the network referral program is created / updated and can be one of the following states. The current state of the network referral program should be exposed via an API.
+
+| Status               | Benefits Enabled | Condition for entry                                       | Condition for exit                                                |
+| -------------------- | ---------------- | --------------------------------------------------------- | ----------------------------------------------------------------- |
+| `STATUS_PROPOSED`    | No               | Governance proposal valid and accepted                    | Governance proposal voting period ends (or proposal is invalid)   |
+| `STATUS_REJECTED`    | No               | Governance vote fails (or is invalid)                     | New governance proposal submitted to the network                 |
```

Comment:
If there is an active programme and a new proposal is rehected or fails, what is the status of the programme? Should strill be active but not sure if that is clear here.

I'd suggest maybe rejected is not a valid state for the programme, and you only have:

**INACTIVE**: there is no current programme and no proposal either still in voting or passed voting for there to be one
**PROPOSED**: there is no current programme and no proposal that has passed voting but at least one proposal is still being voted on
**PENDING**: there is no current programme but a proposal has passed voting (voting has closed and criteria were met) but not yet been enacted
**ACTIVE**: there is currently an active referral programme